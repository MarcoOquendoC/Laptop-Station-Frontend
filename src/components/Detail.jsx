import React, { useState, useEffect, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Link,
  NavLink,
  useLocation,
  useParams,
} from 'react-router-dom';
import { getItemsInfo } from '../redux/Home/home';
import NavPrincipal from './NavPrincipal';

const Detail = () => {
  const params = useParams();
  const { id } = params;
  const location = useLocation();
  const dispatch = useDispatch();

  const handleGetItemsInfo = useCallback(() => {
    dispatch(getItemsInfo());
  }, [dispatch]);

  useEffect(() => {
    handleGetItemsInfo();
  }, [handleGetItemsInfo]);

  const [state, updateState] = useState('');
  const laptops = useSelector((store) => store.items);
  const [laptop] = laptops.filter((laptop) => laptop.id === (Number(id)));

  const handleBack = () => {
    if (location.pathname === '/detail/:id') {
      updateState(true);
    } else {
      updateState(!state);
    }
  };

  return (
    <>
      <NavPrincipal />
      {laptop
        ? (
          <div className="viewfinder">
            <NavLink to="/" onClick={() => handleBack()} className="btn left">Back</NavLink>
            <div className="details">
              <div className="detail_img">
                <img src={laptop.image} className="image-detail" alt="laptop" />
              </div>
              <div className="details_item">
                <h1>{laptop.title}</h1>
                <h4>{laptop.item_model}</h4>
                <h2>{laptop.brand}</h2>
                <h4>{laptop.serial_n}</h4>
                <p>{laptop.description}</p>
                <div className="details_btn">
                  <Link to={`/detail/${id}/reserves`} state={id} key={id} className="details_btn__reserve" data-testid="reserve-link">Reserve</Link>
                  <button className="details_btn__admin" type="button">Edit Item</button>
                  <button className="details_btn__admin" type="button">Delete Item</button>
                </div>
              </div>
            </div>
          </div>
        ) : null}
    </>
  );
};

export default Detail;
