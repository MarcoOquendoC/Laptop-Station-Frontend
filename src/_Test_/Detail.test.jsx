import { getByAltText, getByText } from '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Detail from '../components/Detail';

describe('Testing Detail', () => {
  let details;
  beforeAll(async () => {
    details = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Detail />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  it('should renders correctly', () => {
    expect(details).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(details).toMatchSnapshot();
  });

  it('Should have a link to the reservation page', () => {
    const link = screen.findByTestId('reserve-link');
    waitFor(() => expect(link).toBeInTheDocument());
    waitFor(() => expect(link).toHaveAttribute('href'));
  });

  const mockLaptops = [
    {
      id: 1,
      image_url: 'https://limalocal.com/laptopstation/images/rog_strix.png',
      title: 'ROG Strix SCAR 15',
      item_model: 'G532LW-AZ088T',
      serial_n: '00000_000_0000000',
      description: 'NVIDIA® GeForce RTX™ 4080 Laptop GPU.',
    },
  ];

  jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({ id: '1' }),
  }));

  it('Details of a laptop are correctly displayed', () => {
    jest.spyOn(store, 'getState').mockReturnValue({
      trips: {
        status: 'succeeded',
        trips: mockLaptops,
      },
    });

    waitFor(() => expect(getByAltText('details')).toHaveAttribute('src', 'https://limalocal.com/laptopstation/images/rog_strix.png'));
    waitFor(() => expect(getByText('ROG Strix SCAR 15')).toBeInTheDocument());
    waitFor(() => expect(getByText('G532LW-AZ088T')).toBeInTheDocument());
    waitFor(() => expect(getByText('00000_000_0000000')).toBeInTheDocument());
    waitFor(() => expect(getByText('NVIDIA® GeForce RTX™ 4080 Laptop GPU.')).toBeInTheDocument());
  });
});
