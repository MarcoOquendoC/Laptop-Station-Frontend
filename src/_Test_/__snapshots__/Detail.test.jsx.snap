// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Detail should match the snapshot 1`] = `
<div
  className="principal"
>
  <div
    className="header"
  >
    <div
      className="menu mob"
    >
      <img
        alt="menu"
        src="menu.svg"
      />
    </div>
    <div
      className="menu_logo"
    >
      <img
        alt="logo"
        className="logo"
        src="logo_laptop_station.svg"
      />
    </div>
    <div
      className="menu mob"
    />
  </div>
  <nav
    className="menunav"
  >
    <ul
      className="menunav_ul"
    >
      <li
        className="menu_text"
      >
        <a
          href="/login"
          onClick={[Function]}
        >
          Log in
        </a>
      </li>
      <li
        className="menu_text select"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
      </li>
      <li
        className="menu_text"
      >
        <a
          href="/laptops"
          onClick={[Function]}
        >
          Latops
        </a>
      </li>
      <li
        className="menu_text"
      >
        <a
          href="/reserves"
          onClick={[Function]}
        >
          Reserves
        </a>
      </li>
      <li
        className="menu_text admin"
      >
        <a
          className="a__admin"
          href="/addItem"
          onClick={[Function]}
        >
          Add item
        </a>
      </li>
      <li
        className="menu_text"
      />
    </ul>
  </nav>
</div>
`;
